addons/binding/org.openhab.binding.icloud/src/main/java/org/openhab/binding/icloud/internal/BridgeHandlerFactory.java

```diff
@@ -0,0 +1,88 @@
+/**
+ * Copyright (c) 2010-2017 by the respective copyright holders.
+ *
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ */
+package org.openhab.binding.icloud.internal;
+
+import static org.openhab.binding.icloud.BindingConstants.*;
+
+import java.util.HashMap;
+import java.util.Hashtable;
+import java.util.Map;
+
+import org.eclipse.jdt.annotation.Nullable;
+import org.eclipse.smarthome.core.thing.Bridge;
+import org.eclipse.smarthome.core.thing.Thing;
+import org.eclipse.smarthome.core.thing.ThingTypeUID;
+import org.eclipse.smarthome.core.thing.ThingUID;
+import org.eclipse.smarthome.core.thing.binding.BaseThingHandlerFactory;
+import org.eclipse.smarthome.core.thing.binding.ThingHandler;
+import org.openhab.binding.icloud.discovery.DeviceDiscovery;
+import org.openhab.binding.icloud.handler.BridgeHandler;
+import org.openhab.binding.icloud.handler.DeviceHandler;
+import org.osgi.framework.ServiceRegistration;
+import org.osgi.service.component.annotations.Component;
+
+/**
+ * The {@link BridgeHandlerFactory} is responsible for creating things and thing
+ * handlers.
+ *
+ * @author Patrik Gfeller - Initial contribution
+ */
+@Component(service = BridgeHandlerFactory.class, immediate = true, configurationPid = "binding.icloud")
+public class BridgeHandlerFactory extends BaseThingHandlerFactory {
+    private final Map<ThingUID, @Nullable ServiceRegistration<?>> discoveryServiceRegistrations = new HashMap<>();
+
+    @Override
+    public boolean supportsThingType(ThingTypeUID thingTypeUID) {
+        return SUPPORTED_THING_TYPES_UIDS.contains(thingTypeUID);
+    }
+
+    @Override
+    protected ThingHandler createHandler(Thing thing) {
+        ThingTypeUID thingTypeUID = thing.getThingTypeUID();
+
+        if (thingTypeUID.equals(THING_TYPE_ICLOUD)) {
+            BridgeHandler bridgeHandler = new BridgeHandler((Bridge) thing);
+            registerDeviceDiscoveryService(bridgeHandler);
+            return bridgeHandler;
+        }
+
+        if (thingTypeUID.equals(THING_TYPE_ICLOUDDEVICE)) {
+            return new DeviceHandler(thing);
+        }
+        return null;
+    }
+
+    @Override
+    protected void removeHandler(ThingHandler thingHandler) {
+        if (thingHandler instanceof BridgeHandler) {
+            unregisterDeviceDiscoveryService((BridgeHandler) thingHandler);
+        }
+    }
+
+    private synchronized void registerDeviceDiscoveryService(BridgeHandler bridgeHandler) {
+        DeviceDiscovery discoveryService = new DeviceDiscovery(bridgeHandler);
+        discoveryService.activate();
+        this.discoveryServiceRegistrations.put(bridgeHandler.getThing().getUID(), bundleContext
+                .registerService(DeviceDiscovery.class.getName(), discoveryService, new Hashtable<String, Object>()));

```

Comment:
Just like in the Hue binding, you need to register it as `DiscoveryService.class.getName()`.

Reply:
@kaikreuzer Sorry; I do not understand your comment. What do I need to change?

Reply:
Replace `DeviceDiscovery.class.getName()` by `DiscoveryService.class.getName()` in line 72. The first parameter is the name of the interface under which the service is to be registered. And you want to register it as a DiscoveryService.

Reply:
Thanks :-) ... => Done.